File: frontend/src/lib/score.ts
// Direct port of discovery/score.py so the dashboard's numbers come from the
// same two-axis model as the real pipeline, not arbitrary mock numbers.
import {
  FOOTPRINT_DISQUALIFY,
  OBSCURITY_K,
  STRONG_EXCLUDERS,
  WEIGHTS,
} from "./constants";
import type { FootprintRow, QualityBreakdown, Signals } from "./types";

export function rawDimensions(sig: Signals): QualityBreakdown {
  const rating = sig.rating;
  const reviews = sig.review_count;
  const lastDays = sig.last_activity_days;
  return {
    longevity: sig.longevity_years,
    customer: (rating / 5.0) * Math.log1p(reviews),
    consistency: 1.0 / (1.0 + lastDays / 30.0),
    growth: sig.locations + sig.job_postings,
    revenue: sig.tenders_won + sig.locations,
    validation: Math.max(sig.certified, sig.association_member),
  };
}

function normalize(values: number[]): number[] {
  const lo = Math.min(...values);
  const hi = Math.max(...values);
  if (hi - lo < 1e-9) return values.map(() => 0.5);
  return values.map((v) => (v - lo) / (hi - lo));
}

export function obscurityFromHits(totalHits: number): number {
  return 100.0 / (1.0 + totalHits / OBSCURITY_K);
}

export interface ScoreInput {
  id: number;
  signals: Signals;
  footprint: FootprintRow[];
}

export interface ScoreResult {
  qualityBreakdown: QualityBreakdown;
  qualityContributions: QualityBreakdown;
  quality: number;
  obscurity: number;
  hc_rank: number;
  disqualified: boolean;
  disqualifyReason: string;
  totalFootprintHits: number;
}

// Scores an entire sector cohort at once, since Quality dimensions are
// min-max normalised *within the sector* — exactly like score_sector() in Python.
export function scoreCohort(items: ScoreInput[]): Map<number, ScoreResult> {
  const raws = items.map((i) => rawDimensions(i.signals));
  const dims = Object.keys(WEIGHTS) as (keyof typeof WEIGHTS)[];
  const normed: Record<string, number[]> = {};
  for (const dim of dims) {
    normed[dim] = normalize(raws.map((r) => r[dim]));
  }

  const results = new Map<number, ScoreResult>();
  items.forEach((item, i) => {
    const quality =
      100.0 *
      dims.reduce((sum, dim) => sum + WEIGHTS[dim] * normed[dim][i], 0);

    const totalHits = item.footprint.reduce((s, f) => s + f.hits, 0);
    const obscurity = obscurityFromHits(totalHits);

    const strongHit = item.footprint.find(
      (f) => f.hits > 0 && STRONG_EXCLUDERS.has(f.source)
    );
    let disqualified = false;
    let reason = "";
    if (strongHit) {
      const strongSources = item.footprint
        .filter((f) => f.hits > 0 && STRONG_EXCLUDERS.has(f.source))
        .map((f) => f.source)
        .sort();
      disqualified = true;
      reason = `listed in common database(s): ${strongSources.join(", ")}`;
    } else if (totalHits > FOOTPRINT_DISQUALIFY) {
      disqualified = true;
      reason = `ecosystem footprint too high (${totalHits} hits)`;
    }

    const hc = disqualified
      ? 0
      : Math.sqrt(Math.max(quality, 0) * Math.max(obscurity, 0));

    const contributions = dims.reduce((acc, dim) => {
      acc[dim] = Math.round(WEIGHTS[dim] * normed[dim][i] * 100 * 10) / 10;
      return acc;
    }, {} as QualityBreakdown);

    results.set(item.id, {
      qualityBreakdown: raws[i],
      qualityContributions: contributions,
      quality: Math.round(quality * 10) / 10,
      obscurity: Math.round(obscurity * 10) / 10,
      hc_rank: Math.round(hc * 10) / 10,
      disqualified,
      disqualifyReason: reason,
      totalFootprintHits: totalHits,
    });
  });
  return results;
}

File: discovery/score.py
"""The two-axis scoring model.

Quality  : strength of fundamentals from public signals (0..100)
Obscurity: inverse of ecosystem footprint (0..100)
hc_rank  : geometric mean of the two, so a business must be BOTH excellent AND
           unknown to rank high. A hard gate disqualifies anything too visible.

Quality is built from weighted dimensions. Each dimension is reduced to a raw
number per business, then min-max normalised *within the sector cohort* (you
can only compare a dairy to other dairies, not to a software seller), then
weighted and summed.
"""
from __future__ import annotations

import math

from .models import Score, Sig, now_iso
from .store import Store

# Quality dimension -> (weight, function from latest-signals dict -> raw value).
# Weights sum to 1.0. Tune these as the scoring model matures.
WEIGHTS = {
    "longevity": 0.20,
    "customer": 0.30,
    "consistency": 0.15,
    "growth": 0.15,
    "revenue": 0.10,
    "validation": 0.10,
}


def _g(sig: dict, key: str, default: float = 0.0) -> float:
    row = sig.get(key)
    return float(row["value"]) if row else default


def _raw_dimensions(sig: dict) -> dict[str, float]:
    """Collapse a business's signals into one raw number per Quality dimension."""
    rating = _g(sig, Sig.RATING)            # 0..5
    reviews = _g(sig, Sig.REVIEW_COUNT)
    last_days = _g(sig, Sig.LAST_ACTIVITY_DAYS, default=365)
    return {
        "longevity": _g(sig, Sig.LONGEVITY_YEARS),
        # satisfaction weighted by how many customers it's proven across
        "customer": (rating / 5.0) * math.log1p(reviews),
        # fresher activity => higher; invert days-since-last-activity
        "consistency": 1.0 / (1.0 + last_days / 30.0),
        "growth": _g(sig, Sig.LOCATIONS) + _g(sig, Sig.JOB_POSTINGS),
        "revenue": _g(sig, Sig.TENDERS_WON) + _g(sig, Sig.LOCATIONS),
        "validation": max(_g(sig, Sig.CERTIFIED), _g(sig, Sig.ASSOCIATION_MEMBER)),
    }


def _normalize(values: list[float]) -> list[float]:
    """Min-max to 0..1. Constant column -> all 0.5 (no information)."""
    lo, hi = min(values), max(values)
    if hi - lo < 1e-9:
        return [0.5 for _ in values]
    return [(v - lo) / (hi - lo) for v in values]


# --- Obscurity / disqualification gate --------------------------------------
FOOTPRINT_DISQUALIFY = 8     # total press/funding hits above this => excluded
OBSCURITY_K = 2.0            # softness of the obscurity decay


def _obscurity(total_hits: int) -> float:
    """Map footprint hits -> 0..100. 0 hits = 100 (invisible), decays fast."""
    return 100.0 / (1.0 + total_hits / OBSCURITY_K)


def score_sector(store: Store, sector: str) -> list[Score]:
    rows = store.businesses(sector)
    if not rows:
        return []

    # 1) raw dimensions per business
    raws: list[dict[str, float]] = []
    for b in rows:
        raws.append(_raw_dimensions(store.latest_signals(b["id"])))

    # 2) normalise each dimension across the cohort
    normed = {dim: _normalize([r[dim] for r in raws]) for dim in WEIGHTS}

    # 3) combine
    results: list[Score] = []
    for i, b in enumerate(rows):
        quality = 100.0 * sum(WEIGHTS[dim] * normed[dim][i] for dim in WEIGHTS)
        hits = store.total_footprint(b["id"])
        obscurity = _obscurity(hits)

        disq, reason = 0, ""
        if hits > FOOTPRINT_DISQUALIFY:
            disq, reason = 1, f"ecosystem footprint too high ({hits} hits)"

        hc = 0.0 if disq else math.sqrt(max(quality, 0) * max(obscurity, 0))
        s = Score(business_id=b["id"], quality=round(quality, 1),
                  obscurity=round(obscurity, 1), hc_rank=round(hc, 1),
                  disqualified=disq, reason=reason, computed_at=now_iso())
        store.put_score(s)
        results.append(s)
    return results
